guest/api/clusterevent.go

```diff
@@ -0,0 +1,93 @@
+/*
+   Copyright 2023 The Kubernetes Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package api
+
+import "strconv"
+
+// ActionType is an integer to represent one type of resource change.
+// Different ActionTypes can be bit-wised to compose new semantics.
+type ActionType uint32
+
+// Constants for ActionTypes.
+const (
+	Add    ActionType = 1 << iota // 1
+	Delete                        // 10
+	// UpdateNodeXYZ is only applicable for Node events.
+	UpdateNodeAllocatable // 100
+	UpdateNodeLabel       // 1000
+	UpdateNodeTaint       // 10000
+	UpdateNodeCondition   // 100000
+
+	All ActionType = 1<<iota - 1 // 111111
+
+	// Use the general Update type if you don't either know or care the specific sub-Update type to use.
+	Update = UpdateNodeAllocatable | UpdateNodeLabel | UpdateNodeTaint | UpdateNodeCondition
+)
+
+// GVK is short for group/version/kind, which can uniquely represent a particular API resource.
+type GVK uint32
+
+// Constants for GVKs.
+const (
+	Pod GVK = iota
+	Node
+	PersistentVolume
+	PersistentVolumeClaim
+	PodSchedulingContext
+	ResourceClaim
+	StorageClass
+	CSINode
+	CSIDriver
+	CSIStorageCapacity
+	WildCard
+)
+
+// String implements fmt.Stringer
+func (gvk GVK) String() string {
+	if int(gvk) < len(gvkToString) {
+		return gvkToString[gvk]
+	}
+	return "GVK(" + strconv.Itoa(int(gvk)) + ")"
+}
+
+var gvkToString = [...]string{
+	"Pod",
+	"Node",
+	"PersistentVolume",
+	"PersistentVolumeClaim",
+	"PodSchedulingContext",
+	"ResourceClaim",
+	"storage.k8s.io/StorageClass",
+	"storage.k8s.io/CSINode",
+	"storage.k8s.io/CSIDriver",
+	"storage.k8s.io/CSIStorageCapacity",
+	"*",
+}
+
+// ClusterEvent abstracts how a system resource's state gets changed.
+// Resource represents the standard API resources such as Pod, Node, etc.
+// ActionType denotes the specific change such as Add, Update or Delete.
+type ClusterEvent struct {
+	Resource   GVK
+	ActionType ActionType
+	Label      string
```

Comment:
I took time looking at the logic in the scheduler side around EventToRegister(), and found that no one read this Label from the plugin's EventToRegister().
(asked the reason in the [slack](https://kubernetes.slack.com/archives/C09TP78DV/p1689409183711429) as I'm not familiar with the history of EventToRegister.)
Actually, no plugin in the upstream return event with Label in `EventToRegister()`, so we can probably remove this if we want.

Do you think removing this `Label` would make things better/easier?

Reply:
I would love to remove it, yes. Then the data type is fixed width. I would say yes to removing anything (except ctx and error return 😎 )